CitationStyleLanguagePlugin.inc.php

```diff
@@ -390,14 +517,47 @@ public function getCitation($request, $article, $citationStyle = 'apa', $issue =
                     $currentAuthor->family = htmlspecialchars($author->getLocalizedFamilyName());
                     $currentAuthor->given = htmlspecialchars($author->getLocalizedGivenName());
                 }
-                $citationData->author[] = $currentAuthor;
+                $userGroup = $author->getUserGroup();
+                if (null !== $userGroup) {
+                    switch (true) {
+                        case in_array($userGroup->getId(),$editorsGroup,false):
+                            if (!isset($citationData->editor)) {
+                                $citationData->editor = [];
+                            }
+                            $citationData->editor[] = $currentAuthor;
+                            break;
+                        case in_array($userGroup->getId(),$translatorsGroup,false):
+                            if (!isset($citationData->translator)) {
+                                $citationData->translator = [];
+                            }
+                            $citationData->translator[] = $currentAuthor;
+                            break;
+                        case $chapter && in_array($userGroup->getId(),$authorsGroup,false):
+                            if (!isset($citationData->{'container-author'})) {
+                                $citationData->{'container-author'} = [];
+                            }
+                            $citationData->{'container-author'}[] = $currentAuthor;
+                            break;
```

Comment:
If I remember rightly, I need this for monographs with a preface or a comment from an other author. In case of preface/comment, the chapter author is not author of the book. For CSL the book author has to become 'container-author' at that chapter.
I think it's mostly relevant for translations or critical editions. If you have to publish a new translation of  an ancient text, volume author is the ancient author, translator is clear, but often you find also some comments from translator or an other person to this ancient text.

## Example:
### Citation volume:
Albertus Magnus, De caelo et mundi, ed. by Paul HOSSFELD (Alberti Magni Opera Omnia, Editio Coloniensis, V,1), Münster 1970.

->author: Albertus Magnus
->editor: Paul Hossfeld

### Citation comment:
OPELT, Ilona, Prolegomena, in Albertus Magnus, De caelo et mundi, ed. by Paul HOSSFELD (Alberti Magni Opera Omnia, Editio Coloniensis, V,1), Münster 1970, p. V - XVII.

->chapter author: Ilona Opelt
->container author: Albertus Magnus
->editor: Paul Hossfeld

Reply:
Thanks! I think this might lead to all other chapter authors being identified as `container-author`s. Here the `$currentAuthor` is any author added to the publication. So consider the following case of an edited volume:

- Introduction: Dr Lorem Ipsum
- Chapter 1: Dolor Sitamet
- Chapter 2: Quis Nostrud

The publication authors (`$publication->getData('authors')`) will be Dr Lorem Ipsum, Dolor Sitamet and Quis Nostrud. If I understand the code above correctly, the citation data for Chapter 1 would look like this:

```
[
    'editor' => [Dr Lorem Ipsum],
    'container-author' => [Dolor Sitamet, Quis Nostrud],
]
```

Is that right?

Reply:
Yes, you're right. I will change this.

Reply:
Sorry, that it takes such a long time.  Now, I'm sure here is no need to change, it works like expected. It is only necessery to set the plugin-settings and metadata correct. If "Dolor Sitamet" and "Quis Nostrud" get contributer's role author, it works like you told, but if they get the role chapter-author, it works right, because the line `case $chapter && in_array($userGroupId,$authorsGroups):`. The role-id for chapter-authors shouldn't part of the authorsGroups array you can define in the settings. I'm not sure how to explain this better. Let me show you some screenshots:

plugin-settings:

![2022-02-02 10_28_13-Website Settings _ TestPress](https://user-images.githubusercontent.com/78475910/152130097-bf3fe232-191b-43e8-9f78-93172b35afcb.png)


**correct metadata:**

![2022-02-02 10_46_57-Ipsum et al  _ At vero eos _ TestPress](https://user-images.githubusercontent.com/78475910/152130483-b8ad8f17-9c19-46a2-be15-4dafd43e8cf9.png)


results:

![2022-02-02 10_30_34-At vero eos _ TestPress](https://user-images.githubusercontent.com/78475910/152130636-43231f2f-7db4-4aaa-8911-e41b4365c1e7.png)


![2022-02-02 10_30_14-Introduction _ TestPress](https://user-images.githubusercontent.com/78475910/152130681-0aadb18d-4649-4c2c-bc8b-1e9b3450df21.png)


![2022-02-02 10_30_53-Chapter 1 _ TestPress](https://user-images.githubusercontent.com/78475910/152130699-b42b163d-df60-4993-9a12-78d975232f39.png)

![2022-02-02 10_31_22-Chapter 2 _ TestPress](https://user-images.githubusercontent.com/78475910/152130739-bc0cea70-feaa-48fe-951b-1711521e1150.png)



**wrong metadata:**

![2022-02-02 10_51_21-Ipsum et al  _ At vero eos _ TestPress](https://user-images.githubusercontent.com/78475910/152131135-e2bebaee-b892-4761-93d0-13c1955e8763.png)


results:

![2022-02-02 10_52_09-At vero eos _ TestPress](https://user-images.githubusercontent.com/78475910/152131371-1854c016-9d8c-4a3e-8dfd-ccba7c99fcf3.png)

![2022-02-02 10_53_46-Introduction _ TestPress](https://user-images.githubusercontent.com/78475910/152131592-943c3aa8-341e-4b54-b786-c448b889b88a.png)

![2022-02-02 10_55_18-Chapter 1 _ TestPress](https://user-images.githubusercontent.com/78475910/152131763-d19a3988-8cb2-4144-af9e-c91ac9eff714.png)

I hope it is comprehensible how it works.

Reply:
Thanks for the detailed explanation and sorry for the delay in getting back to you. I understand now how it is supposed to work. I can't yet figure out, reading the code, whether that is actually what happens. The code I'm looking closely at is:

```php
switch (true) {
    
    // ...other cases

    case $chapter && in_array($userGroup->getId(),$authorsGroup,false):
        if (!isset($citationData->{'container-author'})) {
            $citationData->{'container-author'} = [];
        }
        $citationData->{'container-author'}[] = $currentAuthor;
        break;
    case in_array($userGroup->getId(),$authorsGroup,false):
        if (!isset($citationData->author)) {
            $citationData->author = [];
        }
        $citationData->author[] = $currentAuthor;
        break;
    default:
        if (!isset($citationData->author)) {
            $citationData->author = [];
        }
        break;
}
```

The two cases above only rely on `$authorsGroup`, and I don't see where the chapter author groups are being defined here. Does this interact in some way with the line much higher?

```php
case $this->isApplicationOmp() && $chapter:
```

It's not clear to me why this code is broken up into two places or how it interacts to produce the final citation data.

(I am also concerned about the UX of this -- it's not clear to me how a press manager will know to configure their author groups correctly. But we can tackle that after we figure out this coding issue.)

Reply:
You're right, this interact with the code in lines 420 - 434. There I use $chapterAuthors = $chapter->getAuthors(); and only look at the chapter authors groups. At line 433 I add a chapter author to the $citationData->author array. For books and articles I fill this array in line 545 based on $authors = $publication->getData('authors');.

Reason to split this in two parts is the different source of the authors-array. For chapters I need both parts, first to get the chapter authors, second for to get editors, translators and possible volume authors. To get volume authors I need case $chapter && in_array($userGroup->getId(),$authorsGroup,false): .

Reason why this two parts are so far from each other is, that I tried to group code, that is only for chapters or only for books or only for articles or needed for all. But I see now that this makes difficulties to understand what happens. I don't see a perfect solution to write well readable code in this case, but I have an idea to make it better. I could do the second part above the first part and remove case $chapter && in_array($userGroup->getId(),$authorsGroup,false):. In the chapter only part I would first remove authors from $citationData-author to $citationData->{'container-author'} and then fill $citationData-author again.
Do you have a better solution?

Reply:
I think one thing that would help readability is to break code out into protected functions. It's not easy to follow the logic of the switch cases, and to know under what conditions each thing will be called.

I think it's ok to duplicate code sometimes if it makes it more readable, especially if you can encapsulate some of the complex conditional statements into well-named functions. Something like this would be more readable, even if it led to some code duplication:

```php
if ($this->isBookChapter(...)) {
    $citationData = $this->setBookChapterAuthors($citationData, ...);
} elseif ($this->isEditedVolume(...)) {
    $citationData = $this->setEditedVolumeAuthors($citationData, ...);
} elseif ($this->isMonograph(...)) {
    $citationData = $this->setMonographAuthors($citationData, ...);
} elseif ($this->isArticle(...)) {
    $citationData = $this->setArticleAuthors($citationData, ...);
}
```

Reply:
Okay, I will do that.